core/src/main/java/io/micronaut/core/execution/ImperativeExecutionFlowImpl.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.core.execution;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Nullable;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * The imperative execution flow implementation.
+ *
+ * @author Denis Stepanov
+ * @since 4.0.0
+ */
+@Internal
+final class ImperativeExecutionFlowImpl implements ImperativeExecutionFlow<Object> {
```

Comment:
If the interface was modified this could be a record

Reply:
I made all the implementations mutable, so we don't create additional instances.